corehq/messaging/scheduling/forms.py

```diff
@@ -1,22 +1,169 @@
+
+from corehq.apps.style import crispy as hqcrispy
+from corehq.apps.translations.models import StandaloneTranslationDoc
+from corehq.apps.users.models import CommCareUser
 
 
 class MessageRecipientWidget(forms.Widget):
-    def __init__(self, domain, attrs=None, id='message-recipient'):
+    def __init__(self, domain, readonly, attrs=None, id='message-recipient'):
         super(MessageRecipientWidget, self).__init__(attrs)
         self.domain = domain
         self.id = id
         self.query_url = reverse('possible_sms_recipients', args=[self.domain])
+        self.readonly = readonly
 
     def render(self, name, value, attrs=None):
+        initial_value = []
+        if value:
+            for doc_type, doc_id in value:
+                user = CommCareUser.wrap(CommCareUser.get_db().get(doc_id))
+                initial_value.append({"id": doc_id, "name": user.raw_username})
         # TODO populate this with inital data if necessary
         return get_template('scheduling/partials/message_recipient_widget.html').render(Context({
             'id': self.id,
             'name': name,
-            'value': '',
+            'value': 'filler' if initial_value else '',
             'query_url': self.query_url,
-            'initial_data': [],
+            'initial_data': initial_value,
+            'readonly': self.readonly
         }))
+
+
+class MessageRecipientField(MultipleChoiceField):
+    def to_python(self, value):
+        if not value:
+            return []
+        return value.split(',')
+
+    def validate(self, value):
+        # TODO Will need to add more than user ids
+        # TODO batch id verification
+        for user_id in value:
+            CommCareUser.get_db().get(user_id)
+
+
+class MessageForm(Form):
+    schedule_name = CharField(
+        required=True,
+        label=_('Schedule Name'),
+        max_length=1000,
+    )
+    send_frequency = ChoiceField(
+        required=True,
+        label=_('Send'),
+        choices=(
+            ('immediately', _('Immediately')),
+        )
+    )
+    recipients = MessageRecipientField(
+        label=_("Recipient(s)"),
+        help_text=_("Type a username, group name or 'send to all'")
+    )
+    content = ChoiceField(
+        required=True,
+        label=_("Content"),  # seems like a weird label?
+        choices=(
+            ('sms', _('SMS')),
+            # ('email', _('Email')),
+            # ('sms_survey', _('SMS Survey')),
+            # ('ivr_survey', _('IVR Survey')),
+        )
+    )
+    translate = BooleanField(
+        label=_("Translate this message"),
+        required=False
+    )
+
+    def __init__(self, *args, **kwargs):
+        self.domain = kwargs.pop('domain')
+        initial = kwargs.get('initial')
+        readonly = False
+        if initial:
+            # TODO Recipients isn't set properly
+            readonly = (initial.get('send_frequency') == 'immediately')
+            message = initial.get('message', dict)
```

Comment:
should this be `message = initial.get('message', {})` ?

Reply:
https://github.com/dimagi/commcare-hq/pull/17722/commits/eef0c7790f399abc7daf0dccd2171432df17b0cb